### ISSUE TITLE
Go to Definition for Magic Methods

### ISSUE THREAD

**Issue by TheColorRed** [bmewburn/intelephense#57]
I know that this may be very difficult to do, so I guess that this is more of a feedback post.

In laravel and other frameworks (and non-frameworks), magic methods are used for certain features, so I would suggest that it could be possible to define a method/property using PHPDoc.

I don't believe that there is a doc comment for this, so I will make one up called: `@called` in this example.

It would look something like this:

```php
class A extends ClassWithMagic {
    /**
     * @called MyMethod
     */
    public function scopeMyMethod() {}
}
```

Then it could be used something like this:

```php
A::MyMethod();
// or
$a = new A;
$a->MyMethod();
// or
(new $a)->MyMethod();
```

It would know about the method due to the doc comment.

**Comment by hworld** [comment]
There is support for this in phpdoc. It's called `@method`. Documentation: https://github.com/phpDocumentor/fig-standards/blob/master/proposed/phpdoc.md#711-method

Looks like intelephense already supports it and works fine.

Would look like this:

```
/**
 * @method string getString()
 * @method void setInteger(int $integer)
 * @method setString(int $integer)
 */
class Child
{
    <...>
}
```

**Comment by TheColorRed** [comment]
That somewhat works, maybe I am doing it wrong, but it doesn't go to the function, just where the comment is. Is there a way to use it so it goes to the method?

```php
/**
 * @method void DefaultSelect()
 */
class GameCache extends BaseGameModel{
    public function scopeDefaultSelect($query, $filters){
		// Do stuff
	}
}
```

**Comment by hworld** [comment]
Oh, I see. So in the __call you end up funneling it to the scopeDefaultSelect (as an example)? Yeah, I don't think that phpdoc allows you to mark that at least...

**Comment by TheColorRed** [comment]
Yeah, that is a Laravel feature for doing database queries:
https://laravel.com/docs/5.5/eloquent#local-scopes

Other frameworks may do something differently

**Comment by arturock** [comment]
I don't think this could be really helpful as code should not depend on an editor. Adding random (and non-sense for others) phpdoc tags will 'couple' the code to one editor plugin, not even an editor, but a plugin feature, ending on dirty code. I'm in a project where some people use Atom, others Sublime or even PhpStorm. I see no reason to have code with custom stuff so each ones editor works, and I cannot imagine my self tagging a scope with @called for me, tagging whatever method with @forwarded so it works on some random Atom plugin and @scopedMethod for my mate using a Sublime plugin. However, suggesting this tag to the phpDoc team may be helpful, some kind of mix between @method and @see

**Comment by TheColorRed** [comment]
I did submit something after. No reply yet though

**Comment by arturock** [comment]
> I did submit something after.

Cheers mate!

I'd love to have this feature supported on phpDoc blocks

**Comment by bmewburn** [comment]
Good discussion, I don't think introducing a new phpdoc tag is appropriate, I'd rather stick with current standards. @arturock mentions `@see`. I wonder if this would be suitable in itself. Document `_call` with `@see` tags and then go to definition can provide multiple locations. It wouldn't distinguish between `scopeMethodA` and `scopeMethodB` but at least provides useful information rather than just going to the `@method` in the class docblock.

**Comment by bmewburn** [comment]
I think the way to go here would be to use inline `@see` in the `@method` description. 
```php
/**
 * @method void DefaultSelect() {@see GameCache::scopeDefaultSelect}
 */
class GameCache extends BaseGameModel{
    public function scopeDefaultSelect($query, $filters){
		// Do stuff
	}
}
```

**Comment by autumndev** [comment]
would this also work with, as an example, laravel facades:
https://github.com/laravel/framework/blob/5.5/src/Illuminate/Support/Facades/DB.php

the ```@see``` directive is class level and not method level. Currently when using a facade no intellisense is provided.
